File: iteration2/Controllers/SimulationController.java
package iteration2.Controllers;

import iteration2.Models.Course;
import iteration2.Models.Curriculum;
import iteration2.Models.Student;
import iteration2.Models.Advisor;

import java.util.ArrayList;


public class SimulationController extends Controller {
    private ArrayList<Student> students;
    private ArrayList<Advisor> advisors;
    private ArrayList<Curriculum> curriculums;

    private ArrayList<Course> deniedCourses;

     SimulationController(ArrayList<Student> students,
                                ArrayList<Curriculum> curriculums, ArrayList<Advisor> advisors) {
        this.students = students;
        this.curriculums = curriculums;
        this.advisors = advisors;
    }

    public void startSimulation() {
        registerStudentsToClasses();
        exportObjects();
        printErrorMessages();
    }

    private void exportObjects() {
        for (Advisor advisor : advisors) {
            exportJSONFile(advisor);
        }
        for (Student student : students) {
            student.getTranscript().calculateGPA();
            exportJSONFile(student);
        }
        exportJSONFile(this.getError());
    }

    private void printErrorMessages() {
        for (String message : this.getError().getAllErrorMessages()) {
            System.out.println(message);
        }
    }

    public void registerStudentsToClasses() {
        StudentController studentController = new StudentController();
        studentController.setError(this.getError());
        this.deniedCourses = new ArrayList<>();

        int studentsCount = students.size();
        Student student = null;

        for (int i = 0; i < studentsCount; i++) {
            student = students.get(i);
            this.deniedCourses.clear();
            //if student registered 2020 and later first indexed curriculum will be taken consideration
            int studentRegistersDateToCurriculum = student.getRegisterDate() >= 2020 ? 1 : 0;
            int simulationSemester = student.getSemesterNo();
            ArrayList<Course> courses = curriculums.get(studentRegistersDateToCurriculum).getCourses().get(simulationSemester);

            if (courses != null) {
                for (Course course : courses) {
                    if (!studentController.registerToCourse(student, course)) {
                        deniedCourses.add(course);
                    }
                }
            } else {
                System.out.println("courses  is null " + student.getSemesterNo());
            }

        }
    }



    public ArrayList<Advisor> getAdvisors() {
        return this.advisors;
    }
}

File: iteration3/Controllers/SimulationController.py
from typing import List
import iteration3.Controllers.Controller as Controller
import iteration3.Models.Student as Student
import iteration3.Models.Advisor as Advisor
import iteration3.Models.Curriculum as Curriculum
import iteration3.Controllers.StudentController as StudentController


class SimulationController(Controller.Controller):
    def __init__(self, students, advisors, curriculums):
        super().__init__()
        self.__students = students
        self.__curriculums = curriculums
        self.__advisors = advisors
        self.__denied_courses = []

    def start_simulation(self):
        #self.register_students_to_classes()
        self.export_objects()
        # self.print_error_messages()
        pass

    def export_objects(self):
        #for advisor in self.__advisors:
            #self.export_json_file(advisor)
        for student in self.__students:
            #student.transcript.calculate_gpa()
            self.export_json_file(student)
        self.export_json_file(self.error)

    def print_error_messages(self):
        for message in self.error.get_all_error_messages():
            print(message)

    def register_students_to_classes(self):
        student_controller = StudentController()
        student_controller.error = self.error
        self.denied_courses.clear()

        for i in range(len(self.students)):
            student = self.students[i]
            self.denied_courses.clear()
            # if student registered 2020 and later first indexed curriculum will be taken consideration
            student_registers_date_to_curriculum = 1 if student.register_date >= 2020 else 0
            simulation_semester = student.semester_no
            courses = self.curriculums[student_registers_date_to_curriculum].courses[simulation_semester]

            if courses is not None:
                for course in courses:
                    if not student_controller.register_to_course(student, course):
                        self.denied_courses.append(course)
            else:
                print(f"courses  is null {student.semester_no}")
